_posts/2024-01-25-azure-without-secrets-part-2.md

```diff
@@ -0,0 +1,380 @@
+  name: storageAccountName
+  location: location
+  properties: {
+    allowSharedKeyAccess: false
+    // ...
+  }
+}
+```
+
+## App Service to Azure SQL Database
+In this example I'll show you how to use identity-based authentication to access data in Azure SQL Database. Granting access to SQL databases is a bit different compared to other examples. We can't use RBAC, because there are no data plane roles for SQL databases. Instead, we need to create a database user for the managed identity and grant database access to that user.
+
+### Set Entra ID authenticated admin
+First, we need to add an Entra ID authenticated database admin. We need this because only Entra ID logged-in admins can create Entra ID authenticated database users. Adding the admin can be done by adding a separate *adminstrators* resource in Bicep. I added az commands for getting values for the current Azure CLI logged-in user.
+
+```bicep
+resource sqlServerAdminstrators 'Microsoft.Sql/servers/administrators@2021-11-01' = {
+  name: 'ActiveDirectory'
+  parent: sqlServer
+  properties: {
+    administratorType: 'ActiveDirectory'
+    login: 'sampleuser@company.tld' // az ad signed-in-user show --query userPrincipalName
+    sid: 'aaaaaaaa-aaaa-aaaa-aaaa-aaaaaaaaaaaa' // az ad signed-in-user show --query id
+    tenantId: 'bbbbbbbb-bbbb-bbbb-bbbb-bbbbbbbbbbbb' // ad account show --query tenantId
+  }
+}
+```
+
+### Create SQL user and grant access rights
+Next, we need to create the database user and grant it access rights. Here are the steps to do that:
+
+1. Connect to the SQL database (not master) with the Entra ID authenticated admin.
+2. Create a user for the managed identity's service principal using the following command (notice that the brackets are required): `CREATE USER [app-service-name] FROM EXTERNAL PROVIDER`
+3. Grants access rights to the newly created user. For example, to grant read and write access, run the following commands: `ALTER ROLE db_datareader ADD MEMBER [app-service-name]` and `ALTER ROLE db_datawriter ADD MEMBER [app-service-name]`
+
+More detailed instructions can be found [here](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/tutorial-windows-vm-access-sql?source=recommendations#create-contained-user).
+
+### Code changes
+#### Add NuGet package
+Entity Framework Core requires a database engine-specific client library, with Azure SQL Database we need `Microsoft.EntityFrameworkCore.SqlServer`. Despite the name, it also supports Azure SQL.
+```shell
+dotnet add package Microsoft.EntityFrameworkCore.SqlServer
+```
+
+#### Add connection strings to the configuration
+Similar to the storage account case, we need to store the database connection string somewhere. Once again, I put it in two places; `ConnectionStrings` in `appsettings.Development.json` for localhost development, and for the App Service, I put it in the App Service's connection strings using Bicep.
```

Comment:
Tähän vois laittaa lisähuomion, että koska meillä ei ole salasanoja connection stringissä, niin tämän voi nyt tehdä näin. Salasanojen kanssa täytyisi käyttää user secrettejä.
https://learn.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-8.0&tabs=windows